exonum-java-binding-common/src/main/java/com/exonum/binding/common/message/TransactionMessage.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright 2018 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.exonum.binding.common.message;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.exonum.binding.common.crypto.CryptoFunction;
+import com.exonum.binding.common.crypto.CryptoFunctions;
+import com.exonum.binding.common.crypto.KeyPair;
+import com.exonum.binding.common.crypto.PublicKey;
+import com.exonum.binding.common.hash.HashCode;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+/**
+ * An Exonum transaction message.
+ */
+public interface TransactionMessage {
+  int AUTHOR_PUBLIC_KEY_OFFSET = 0;
+  int CLS_OFFSET = 32;
+  int TAG_OFFSET = 33;
+  int SERVICE_ID_OFFSET = 34;
+  int TRANSACTION_ID_OFFSET = 36;
+  int PAYLOAD_OFFSET = 38;
+
+  int AUTHOR_PUBLIC_KEY_SIZE = 32;
+  int SIGNATURE_SIZE = 64;
+  int MIN_MESSAGE_SIZE = PAYLOAD_OFFSET + SIGNATURE_SIZE;
+
+  /**
+   * Returns a public key of the author of the transaction message.
+   */
+  PublicKey getAuthor();
+
+  /**
+   * Returns the identifier of the service this message belongs to.
+   */
+  short getServiceId();
+
+  /**
+   * Returns transaction identifier which is unique within the service.
+   */
+  short getTransactionId();
+
+  /**
+   * Returns the transaction message body.
+   */
+  byte[] getPayload();
+
+  /**
+   * Returns the transaction message hash.
+   */
+  HashCode hash();
+
+  /**
+   * Returns the <a href="https://ed25519.cr.yp.to/">Ed25519</a> signature
+   * over this binary message.
+   *
+   * <p>The signature is <strong>not</strong> guaranteed to be valid and must be verified against
+   * the signer’s public key.
+   *
+   * @see CryptoFunctions#ed25519()
+   */
+  byte[] getSignature();
+
+  /**
+   * Returns the transaction message in binary format.
+   */
+  byte[] toBytes();
+
+  /**
+   * Returns the transaction message size in bytes.
+   */
+  int size();
+
+  /**
+   * Create a new builder for the transaction message.
+   */
+  static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * Creates the transaction message from the given bytes array.
+   */
+  static TransactionMessage fromBytes(byte[] bytes) {
+    return new BinaryTransactionMessage(bytes);
+  }
+
+  /**
+   * Creates the transaction message from the given bytes buffer.
+   */
+  static TransactionMessage fromBuffer(ByteBuffer buffer) {
+    return new BinaryTransactionMessage(buffer.array());
```

Comment:
This is broken for the reasons described earlier. Please pass the buffer to the message, so that it works with:
* direct,
* readonly
* with any marks set (e.g., pos=10, lim=128)

byte buffers.